assets/blocks/quiz/quiz-block/use-has-questions.js

```diff
@@ -16,28 +16,39 @@ export const useHasQuestions = ( clientId ) => {
 			.filter( ( block ) => !! block.attributes.title )
 	);
 
+	const { editPost } = useDispatch( 'core/editor' );
+	const setQuizHasQuestions = useCallback(
+		( on ) => {
+			const { _quiz_has_questions: savedValue } =
+				selectData( 'core/editor' ).getCurrentPostAttribute( 'meta' ) ||
+				{};
+			// Don't send an update to null if the meta is already unset.
+			const emptyValue = savedValue ? null : undefined;
+			return editPost( {
+				meta: { _quiz_has_questions: on ? 1 : emptyValue },
+			} );
+		},
+		[ editPost ]
+	);
+
 	// Monitor for valid questions.
 	useEffect( () => {
 		const { _quiz_has_questions: quizHasQuestions } =
 			selectData( 'core/editor' ).getEditedPostAttribute( 'meta' ) || {};
 
-		if ( questionBlocks.length ) {
-			dispatch( 'core/editor' ).editPost( {
-				meta: { _quiz_has_questions: 1 },
-			} );
-		} else if ( quizHasQuestions ) {
-			dispatch( 'core/editor' ).editPost( {
-				meta: { _quiz_has_questions: null },
-			} );
+		if ( ! quizHasQuestions && questionBlocks.length ) {
+			setQuizHasQuestions( true );
+		}
+		if ( quizHasQuestions && ! questionBlocks.length ) {
+			setQuizHasQuestions( false );
 		}
-	}, [ questionBlocks ] );
+	}, [ questionBlocks, setQuizHasQuestions ] );
 
 	// Monitor for quiz block removal.
 	useEffect( () => {
 		return () => {
-			dispatch( 'core/editor' ).editPost( {
-				meta: { _quiz_has_questions: null },
-			} );
+			setQuizHasQuestions( false );
 		};
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
If I didn't miss anything, I think `setQuizHasQuestions` as dependency here would not make any harm. Because `setQuizHasQuestions` has only `editPost` as dependency, and `editPost` never changes. I'm suggesting it because it could avoid issues in a possible future maintenance on the `setQuizHasQuestions`, introducing other dependencies.

Reply:
👍 32b1714 (Though it has an extra dependency now, but seems to be working okay)